types/state/stateTxn.go

```diff
@@ -0,0 +1,334 @@
+package state
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/AccumulateNetwork/accumulate/internal/logging"
+	"github.com/AccumulateNetwork/accumulate/smt/common"
+	"github.com/AccumulateNetwork/accumulate/smt/managed"
+	"github.com/AccumulateNetwork/accumulate/smt/storage"
+	"github.com/AccumulateNetwork/accumulate/smt/storage/database"
+	"github.com/AccumulateNetwork/accumulate/types"
+	"sort"
+	"sync"
+)
+
+type DBTransaction struct {
+	state *StateDB
+}
+
+//func (tx *DBTransactional) init(debug bool) (err error) {
+//
+//	tx.debug = debug
+//	tx.updates = make(map[types.Bytes32]*blockUpdates)
+//	tx.writes = map[storage.Key][]byte{}
+//	tx.transactions.reset()
+//
+//	tx.bpt = pmt.NewBPTManager(tx.db)
+//	managed.NewMerkleManager(tx.db, markPower)
+//
+//	ent, err := tx.GetPersistentEntry(blockIndexKey[:], false)
+//	if err == nil {
+//		tx.blockIndex, _ = common.BytesInt64(ent.Entry)
+//	} else if !errors.Is(err, storage.ErrNotFound) {
+//		return err
+//	}
+//
+//	return nil
+//}
+
+func (s *StateDB) Begin() *DBTransaction {
+	dbTx := &DBTransaction{
+		state: s,
+	}
+	return dbTx
+}
+
+//GetPersistentEntry will pull the data from the database for the StateEntries bucket.
+func (tx *DBTransaction) GetPersistentEntry(chainId []byte, verify bool) (*Object, error) {
+	_ = verify
+	tx.state.Sync()
+
+	if tx.state.db == nil {
+		return nil, fmt.Errorf("database has not been initialized")
+	}
+
+	data, err := tx.state.db.Key("StateEntries", chainId).Get()
+	if errors.Is(err, storage.ErrNotFound) {
+		return nil, fmt.Errorf("%w: no state defined for %X", storage.ErrNotFound, chainId)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get state entry %X: %v", chainId, err)
+	}
+
+	ret := &Object{}
+	err = ret.UnmarshalBinary(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal state for %x", chainId)
+	}
+	//if verify {
+	//todo: generate and verify data to make sure the state matches what is in the patricia trie
+	//}
+	return ret, nil
+}
+
+func (tx *DBTransaction) AddTransaction(chainId *types.Bytes32, txId types.Bytes, txPending, txAccepted *Object) error {
```

Comment:
Instead of adding a copy of `AddTransaction`, change the receiver of the original.